microprofile/graphql/server/src/main/java/io/helidon/microprofile/graphql/server/GraphQlCdiExtension.java

```diff
@@ -0,0 +1,161 @@
+import javax.enterprise.inject.spi.DeploymentException;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.inject.spi.ProcessAnnotatedType;
+import javax.enterprise.inject.spi.ProcessManagedBean;
+import javax.enterprise.inject.spi.WithAnnotations;
+
+import io.helidon.graphql.server.GraphQlSupport;
+import io.helidon.graphql.server.InvocationHandler;
+import io.helidon.microprofile.server.ServerCdiExtension;
+import io.helidon.webserver.Routing;
+
+import graphql.schema.GraphQLSchema;
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.eclipse.microprofile.graphql.ConfigKey;
+import org.eclipse.microprofile.graphql.GraphQLApi;
+import org.eclipse.microprofile.graphql.Input;
+import org.eclipse.microprofile.graphql.Interface;
+import org.eclipse.microprofile.graphql.Type;
+
+import static javax.interceptor.Interceptor.Priority.LIBRARY_BEFORE;
+
+/**
+ * A CDI {@link Extension} to collect the classes that are of interest to Microprofile GraphQL.
+ */
+public class GraphQlCdiExtension implements Extension {
+    private static final Logger LOGGER = Logger.getLogger(GraphQlCdiExtension.class.getName());
+
+    /**
+     * The {@link List} of collected API's.
+     */
+    private final Set<Class<?>> candidateApis = new HashSet<>();
+    private final Set<Class<?>> collectedApis = new HashSet<>();
+
+    /**
+     * Collect the classes that have the following Microprofile GraphQL annotations.
+     *
+     * @param processAnnotatedType annotation types to process
+     */
+    void collectCandidateApis(@Observes @WithAnnotations(GraphQLApi.class) ProcessAnnotatedType<?> processAnnotatedType) {
+        Class<?> javaClass = processAnnotatedType.getAnnotatedType().getJavaClass();
+        this.candidateApis.add(javaClass);
+        if (javaClass.isInterface()) {
+            collectedApis.add(javaClass);
+        }
+    }
+
+    void collectApis(@Observes @WithAnnotations({Type.class, Input.class,
+                                                        Interface.class}) ProcessAnnotatedType<?> processAnnotatedType) {
+        // these are directly added
+        this.collectedApis.add(processAnnotatedType.getAnnotatedType().getJavaClass());
+    }
+
+    void collectNonVetoed(@Observes ProcessManagedBean<?> event) {
+        AnnotatedType<?> type = event.getAnnotatedBeanClass();
+        Class<?> clazz = type.getJavaClass();
+
+        if (candidateApis.remove(clazz)) {
+            collectedApis.add(clazz);
+        }
+    }
+
+    void addGraphQlBeans(@Observes BeforeBeanDiscovery event) {
+        event.addAnnotatedType(GraphQlBean.class, GraphQlBean.class.getName())
+                .add(ApplicationScoped.Literal.INSTANCE);
+    }
+
+    void clearCandidates(@Observes AfterBeanDiscovery event) {
+        candidateApis.clear();
+    }
+
+    void registerWithWebServer(@Observes @Priority(LIBRARY_BEFORE + 9) @Initialized(ApplicationScoped.class) Object event,
```

Comment:
I am not sure why? The runtime start is really for native-image specific handling. In this case, we just want to register with webserver routing. 
So I think the event is OK. 
Do you think this could be an issue?

Reply:
I just wanted this to be double-checked, which it has been now. Should be OK.